### PR TITLE
Kitsu custom comment template

### PR DESCRIPTION
**Pull request by EmberLightVFX**
## Changelog Description
Kitsu allows to write markdown in its comment field. This can be something very powerful to deliver dynamic comments with the help the data from the instance.
This feature is defaults to off so the admin have to manually set up the comment field the way they want.
I have added a basic example on how the comment can look like as the comment-fields default value.

To this I want to add some documentation also but that's on its way when the code itself looks good for the reviewers.

## Additional info
The parsing of the code goes through my custom safe_format() function to catch instances where the key might not exist in the instance.data. If it doesn't exist a warning will be printed in the logs and the returning value will be None, resulting in an empty string.
(Fun fact: chatGPT made the function as I was way to tired to figure it out myself when coding that part)

## Testing notes:
1. Enable Custom Comment Template in settings
2. Publish a shot
3. Check the beautiful comment in Kitsu

### CODE REVIEW

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (Tilix4) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:23]: missing docstring

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (Tilix4) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:24]: Must be outside of the function. With docstring

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (Tilix4) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:48]: What if it is a `review`?

**Review by Tilix4 — CHANGES_REQUESTED** [review]

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:48]: Yeah it should be review as that's what should be uploaded to Kitsu. Fixed

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:24]: Is that really a most?
Currently the `replace_missing()` function access the `safe_format()` function's `**kwargs`
How would `re.sub(pattern, replace_missing, msg)` pass the `**kwargs`? As the `msg` needs to be a string if reading the re.sub's docstring.

Any suggestions?

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (BigRoy) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:39]: Technically the value in `instance.data` could be `None` for something and be valid? :)

```suggestion
            key = match.group(1)
            if key not in kwargs:
                self.log.warning(
                    "Key `{}` was not found in instance.data "
                    "and will be rendered as `` in the comment".format(
                        key
                    )
                )
                return ""
            else:
                return str(kwargs[key])
```

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (BigRoy) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:28]: > Currently the replace_missing() function access the safe_format() function's **kwargs
> How would re.sub(pattern, replace_missing, msg) pass the **kwargs?

Yes, I'd say so.

> Is that really a must?

However, for me personally I'm not really to bothered by it being inline - it's like a lambda function but then PEP8 defined as a function. It seems you both are mostly in the Kitsu code so I'll leave it up to you. Adding a docstring I'd recommend however.

File: openpype/settings/entities/schemas/projects_schema/schema_project_kitsu.json
Comment (BigRoy) [inline openpype/settings/entities/schemas/projects_schema/schema_project_kitsu.json:70]: Technically it's not using data from instance's anatomy but the publishing instance data. The anatomy data would be `instance.data["anatomyData"]` or something like that, right?

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (BigRoy) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:23]:
```suggestion
    custom_comment_template = {
        "enabled": False,
        "comment_template": "{comment}",
    }
```

File: openpype/settings/entities/schemas/projects_schema/schema_project_kitsu.json
Comment (BigRoy) [inline openpype/settings/entities/schemas/projects_schema/schema_project_kitsu.json:60]: Looking at other keys it feels a bit funky that this key is UpperCamelCase, no?

File: openpype/settings/defaults/project_settings/kitsu.json
Comment (BigRoy) [inline openpype/settings/defaults/project_settings/kitsu.json:11]:
```suggestion
            "custom_comment_template": {
```

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:28]: My bad, I don't know how, but I missed the use of `kwargs`... Keep it inside, but add docstring.

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:39]: Wops 😄

File: openpype/settings/entities/schemas/projects_schema/schema_project_kitsu.json
Comment (EmberLightVFX) [reply openpype/settings/entities/schemas/projects_schema/schema_project_kitsu.json:70]: Yeah that is more technically correct, the best type of correct :)

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:32]: True, that looks better. I'm so used at writing MD lately so I defaulted to `` 😄

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:49]: The instance might have the family `kitsu` but isn't a review. In that instance it would still be processed in the plugin.

Comment (BigRoy) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:49]:
```suggestion
            if "review" not in instance.data["families"]:
```
If @Tilix4's suggestion is wrong then likely it should still have to be this if we need to filter. :)

But note that the context has all instances, so also the workfile instance which is likely not marked as `review` but we might still want to integrate a note to kitsu about it? Or are notes ONLY for reviewables?

Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:49]: I found what @Tilix4 wrote on Discord very logical:

> in our current Blender refactor, we basically dropped it because of the inconsistent behaviour you describe. IMO, you publish a review only if you have a review instance in your scene, otherwise it is too hard to track what's going on.

Else the end user would see something like this (is not using this PR to get extra information per comment):
image

So I would say only reviews would be uploaded to Kitsu

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:78]: I don't really see the difference between this and what I'm doing currently. Would this function live inside the process function? I rather have it outside to closer go with `A function does one thing`

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:78]: Yes sure, that was my point, replace `safe_format` with this one, and keep it outside process. The idea is that `safe_format` is dedicated to this, therefore I don't think we need a function for it, while `get_publish_comment` makes sense and becomes overridable by giving its own function.

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py]: I agree with @EmberLightVFX who agrees with me.

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (BigRoy) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:78]: I think you're saying that if we change it to for example this:

```python
def format_publish_comment(self, instance):
    """Get publish comment build from template."""
    
    def replace_missing_key(match):
            key = match.group(1)
            if key not in instance.data:
                self.log.warning(
                    "Key "{}" was not found in instance.data "
                    "and will be rendered as "" in the comment".format(key)
                )
                return ""
            else:
                return str(instance.data[key])

    template = self.custom_comment_template["comment_template"]
    pattern = r"\{([^}]*)\}"
    return re.sub(pattern, replace_missing_key, template)
```

The code in process could then just do:
```py
publish_comment = self.format_publish_comment(instance)
```

And as such if someone would like to basically redefine completely how it's formatted they could just inherit and override only `format_publish_comment` without requiring to use the requiring to use the comment template, and potentially even take other things from the instance, like formatting with data from `instance.context` if really needed.

I haven't seen anyone inherit from other Pyblish plug-ins without that being a "base class" meant solely for inheriting from - but I suppose this does open the door to turn it into that.

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:78]: That's what I meant, but it might not be a strong requirement. My point was the `safe_format` function is only used once, so maybe not necessary to be a function, and thinking about a case we would need it to be a function, I proposed this.

Short story: `safe_format` adds confusion IMO and could be more simple.

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:36]: This looks like unfinished sentence
```suggestion
                    "and will be rendered as an empty string"
```

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (BigRoy) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:83]: I don't think they'd want to skip if there's no comment because I think this is also the thing that actually adds the reviewable image/video as an attachment to Kitsu. (I think it is, not sure)

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:83]: Oh, really?

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:83]: It is changin status, I see....

Comment (EmberLightVFX) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:83]: Oh it can be? I thought it was one comment per publish and not per instance? How would you place different comments per instance in the new publisher?

And if there is no comment we should still create a comment as it's what the review image/video later will get attached to :) As the script checks the instance for `review` in families only instance that should be published with a review will get a comment

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:83]: There is an [Instance collector](https://github.com/ynput/OpenPype/blob/develop/openpype/plugins/publish/collect_comment.py#L31) which gives option to fill comment per instance. If user won't fill it, the context comment is used. To avoid possible mismatch during publishing the collector will also fill final value to instance data to avoid these `instance.data.get("comment", instance.context.data.get("comment"))`.

The collector must be enabled and families must be defined.

NOTE: I just found out that the collector inherit from `ContextPlugin` instead of `InstancePlugin`, but it works (pyblish magic).

Comment (BigRoy) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py]: @iLLiCiTiT I thought that wasn't the case for [`InstancePlugin` and `ContextPlugin` because they inherit from the `ExplicitMetaClass`](https://github.com/pyblish/pyblish-base/blob/master/pyblish/plugin.py#L320-L356). But following the logic it does seem it might still pass as long as it has either `context` or `instance` as an argument. This might actually be a bug in Pyblish since those two plug-ins were intended to be explicit.

**Comment by EmberLightVFX** [comment]
The one thing missing is documentation but I'm waiting for https://github.com/ynput/OpenPype/pull/4537 to be merged as it already have done some documentation for its feature.
But I can create another PR with just the documentation after this one gets merged :)

**Review by mkolar — APPROVED** [review]

**Comment by mkolar** [comment]
Considering this is affecting some productions right now. let's merge as is to develop and document in #4537

**Pull request merged**
